src/router.ts

```diff
@@ -28,31 +28,36 @@ export interface Router<T> {
 
 export const IS_PATTERN = /[*:{}+?()]/;
 
+function isURLPatternParam(value: unknown): value is string {
+  return typeof value === "string" && value !== "/*" && IS_PATTERN.test(value);
+}
+
 export class UrlPatternRouter<T> implements Router<T> {
   readonly _routes: Route<T>[] = [];
-  readonly _middlewares: T[] = [];
-
-  addMiddleware(fn: T): void {
-    this._middlewares.push(fn);
+  readonly _middlewares: Route<T>[] = [];
+
+  addMiddleware(pathname: string | URLPattern, handler: T): void {
+    this._middlewares.push({
+      path: isURLPatternParam(pathname)
+        ? new URLPattern({ pathname })
+        : pathname,
+      handlers: [handler],
+      method: "ALL",
+    });
   }
 
-  add(method: Method | "ALL", pathname: string | URLPattern, handlers: T[]) {
-    if (
-      typeof pathname === "string" && pathname !== "/*" &&
-      IS_PATTERN.test(pathname)
-    ) {
-      this._routes.push({
-        path: new URLPattern({ pathname }),
-        handlers,
-        method,
-      });
-    } else {
-      this._routes.push({
-        path: pathname,
-        handlers,
-        method,
-      });
-    }
+  add(
+    method: Method | "ALL",
+    pathname: string | URLPattern,
+    handlers: T[],
+  ): void {
```

Comment:
```suggestion
  ) {
```
Nit: the Deno team consider defining `void` and `Promise<void>` as explicit return types as needless boilerplate.